resteasy/src/main/java/com/linecorp/armeria/server/resteasy/ResteasyAsynchronousExecutionContextImpl.java

```diff
@@ -0,0 +1,150 @@
+     * @throws IllegalStateException if it was previously suspended.
+     */
+    @Override
+    public ResteasyAsynchronousResponse suspend(long millis) {
+        return suspend(millis, TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Suspends client connection for asynchronous request processing.
+     * @param time an amount of time to suspend client connection for, expressed in provided units.
+     * @param unit an {@link TimeUnit} for {@code time} parameter.
+     * @throws IllegalStateException if it was previously suspended.
+     */
+    @Override
+    public ResteasyAsynchronousResponse suspend(long time, TimeUnit unit) {
+        if (wasSuspended) {
+            throw new IllegalStateException(Messages.MESSAGES.alreadySuspended());
+        }
+        wasSuspended = true;
+        return asyncResponse;
+    }
+
+    @Override
+    public void complete() {
+        if (wasSuspended) {
+            asyncResponse.complete();
+        }
+    }
+
+    @Override
+    public CompletionStage<Void> executeAsyncIo(CompletionStage<Void> f) {
+        // check if this CF is already resolved
+        final CompletableFuture<Void> ret = f.toCompletableFuture();
+        // if it's not resolved, we may need to suspend
+        if (!ret.isDone() && !isSuspended()) {
+            suspend();
+        }
+        return ret;
+    }
+
+    @Override
+    public CompletionStage<Void> executeBlockingIo(RunnableWithException f,
+                                                   boolean hasInterceptors) {
+        if (!isIoThread()) {
+            // we're blocking
+            try {
+                f.run();
+            } catch (Exception e) {
+                final CompletableFuture<Void> ret = new CompletableFuture<>();
+                ret.completeExceptionally(e);
+                return ret;
+            }
+            return CompletableFuture.completedFuture(null);
+        } else if (!hasInterceptors) {
+            final Map<Class<?>, Object> context = ResteasyContext.getContextDataMap();
+            // turn any sync request into async
+            if (!isSuspended()) {
+                suspend();
+            }
+            return CompletableFuture.runAsync(() -> {
+                try (CloseableContext newContext = ResteasyContext.addCloseableContextDataLevel(context)) {
+                    f.run();
+                } catch (RuntimeException e) {
+                    throw e;
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+            });
+        } else {
+            final CompletableFuture<Void> ret = new CompletableFuture<>();
+            ret.completeExceptionally(new RuntimeException(
+                    "Cannot use blocking IO with interceptors when we're on the IO thread"));
+            return ret;
+        }
+    }
+
+    private static boolean isIoThread() {
+        return Thread.currentThread() instanceof FastThreadLocalThread;
```

Comment:
We use `NonBlocking` for checking it.
https://github.com/line/armeria/blob/3d30b065ee1a1542f54ffe3a03dea6b4cc36c959/core/src/main/java/com/linecorp/armeria/client/ClientFactory.java#L112

Reply:
I took it from RESTEasy Netty binding (see [NettyHttpRequest](https://github.com/resteasy/Resteasy/blob/master/server-adapters/resteasy-netty4/src/main/java/org/jboss/resteasy/plugins/server/netty/NettyHttpRequest.java) as an example) `FastThreadLocalThread` and `NonBlocking` don't seem to be interrelated. Could you explain how `NonBlocking` equivalent to Netty's `FastThreadLocalThread`? Is it applicable to the server context?